### ISSUE TITLE
Is it ok to use 'MASS BASE SI' unit for mixture in the REFPROPdll ?

### ISSUE THREAD

**Issue by LL8848** [usnistgov/REFPROP-wrappers#288]
I came across a similar problem as reported in the issue #267 that REFPROPdll may return a wrong results for mixture if given the unit of MASS BASE SI and mass fractions for the composition. Regarding this issue @ianhbell suggested that we should always use mole fractions.  I created a separate issue here because I found it interesting that REFPROPdll works for mass fractions under some circumstances (see the codes below).  I'm wondering does @ianhbell mean mass fraction is discouraged not not allowed in the REFPROP wrapper?  Is it safe to use MASS BASED SI units ?  (I believe many users prefer to MASS BASED SI units while using REFPROP)

### Steps to Reproduce

``` python
import os; os.environ['RPPREFIX'] = r'C:/Program Files (x86)/REFPROP'
from ctREFPROP.ctREFPROP import REFPROPFunctionLibrary

RP = REFPROPFunctionLibrary(os.environ['RPPREFIX'])
MASS_BASE_SI = RP.GETENUMdll(0, "MASS BASE SI").iEnum
```
If I use a predefined mixture, say R410A, it works perfectly:
``` python
RP.REFPROPdll('R410A',"TQ","P;XMASS",MASS_BASE_SI,1,0,273,0,[0]).Output[:3]
```
**Output:** array('d', [796879.3391703876, 0.4999999999999998, 0.5000000000000001])
This is consistent with the results from GUI.

Then if I specify the fluids explicitly:
``` python 
RP.REFPROPdll('R32 * R125',"TQ","P;XMASS",MASS_BASE_SI,1,0,273,0,[0.5,0.5]).Output[:3]
```
**Output**: array('d', [771732.4395704645, 0.3023853006241376, 0.6976146993758624]). 
The result is supposed to be the same as the R410A but it's different, and output mass fractions are not the same as what I specified.  It seems that the dll here took the composition array as mole fraction, though I specified iUnit = MASS BASED SI and set iMass=1.

What is interesting is that when I ran the above command again in the same code, the result somehow became correct:
``` python 
RP.REFPROPdll('R32 * R125',"TQ","P;XMASS",MASS_BASE_SI,1,0,273,0,[0.5,0.5]).Output[:3]
```
**Output**: array('d', [796879.3391704276, 0.5000000000000001, 0.5])

I found the following codes seem to avoid the errors above.
``` python
RP.SETFLUIDSdll('R32 * R125')
RP.REFPROPdll('',"TQ","P;XMASS",MASS_BASE_SI,1,0,273,0,[0.5,0.5]).Output[:3]
```
**Output**:  array('d', [796879.3391704276, 0.5000000000000001, 0.5])

### Versions

**REFPROP Version:** REFPROP 10.0
**Operating System and Version:** Windows 10
**Access Method:** Python

### Additional Information
Notebook:
![image](https://user-images.githubusercontent.com/42684722/82158631-162f5980-9857-11ea-9740-af59b6f23a9d.png)

**Comment by ianhbell** [comment]
You are right, and this is a perfect embodiment of the bug; the bug has to do with how caching of compositions is handled internally.  All the internal routines use molar composition, which is why I recommend to use molar compositions (only) in REFPROPdll.

**Comment by LL8848** [comment]
I see.  Thank you, Ian! 

- Lingnan

**Comment by EricLemmon** [comment]
Can you send me a personal email (eric.lemmon@nist.gov)?

**Comment by EricLemmon** [comment]
I apologize if you already sent me your email address, I am not sure if you sent it and if I sent you the patch.  Could you please email me again and let me know.

**Comment by LL8848** [comment]
Hi @EricLemmon , 
Just saw your message and I'm so sorry for missing the previous one.  I've sent an email to you.
Thanks for following up the issue.  Regarding that issue, I think I’ve followed  @ianhbell 's advice to use molar composition instead of mass, which works well.  But it would still be great to have a patch that solves this bug.

Cheers,
Lingnan
> 
> 
> I apologize if you already sent me your email address, I am not sure if you sent it and if I sent you the patch. Could you please email me again and let me know.

**Comment by GentleVincent** [comment]
> You are right, and this is a perfect embodiment of the bug; the bug has to do with how caching of compositions is handled internally. All the internal routines use molar composition, which is why I recommend to use molar compositions (only) in REFPROPdll.

Hello, I need to calculate the properties of prededined mixed R448A，the known input value is "TQ"，where the "Q" is Qmass，as you replied above，Qmole is recommend，how to convert the value Qmass to Qmole？thanks!

**Comment by ianhbell** [comment]
Do you mean Qmass as the mass-based vapor quality? Or that you are providing mass fractions?

**Comment by GentleVincent** [comment]
> Do you mean Qmass as the mass-based vapor quality? Or that you are providing mass fractions?

vapor mass to total mass in two phase zone.

**Comment by ianhbell** [comment]
The conversion can be re-derived. The mass and molar qualities are defined by 
$q_{\rm mass} = \frac{m_{\rm v}}{m_{\rm tot}}$
$q_{\rm molar} = \frac{n_{\rm v}}{n_{\rm tot}}$
with $n$ a number of moles and $m$ a mass. The respective number of moles can be obtained from
$n_{\rm tot} = \frac{m_{\rm tot}}{M_{\rm tot}}$
$n_{\rm v} = \frac{m_{\rm v}}{M_{\rm v}}$
with $M$ the molar mass (weighted by the mole fractions in the respective calculation) so

$q_{\rm molar} = \dfrac{\frac{m_{\rm v}}{M_{\rm v}}}{\frac{m_{\rm tot}}{M_{\rm tot}}} = \frac{M_{\rm tot}}{M_{\rm v}}\frac{m_{\rm v}}{m_{\rm tot}}$

$\boxed{q_{\rm molar} = q_{\rm mass}\frac{M_{\rm tot}}{M_{\rm v}}}$

**Comment by GentleVincent** [comment]
> The conversion can be re-derived. The mass and molar qualities are defined by qmass=mvmtot qmolar=nvntot with n a number of moles and m a mass. The respective number of moles can be obtained from ntot=mtotMtot nv=mvMv with M the molar mass (weighted by the mole fractions in the respective calculation) so
> 
> qmolar=mvMvmtotMtot=MtotMvmvmtot
> 
> qmolar=qmassMtotMv

Thank you very much!
